src/mozilla_sec_eia/ex_21/inference.py

```diff
@@ -0,0 +1,366 @@
+            subsidiary, loc, own_per.
+    """
+    dataset = create_inference_dataset(
+        pdfs_dir=pdfs_dir, labeled_json_dir=labeled_json_dir, has_labels=has_labels
+    )
+    if dataset_ind:
+        dataset = dataset.select(dataset_ind)
+
+    # TODO: figure out device argument
+    pipe = pipeline(
+        "token-classification",
+        model=model,
+        tokenizer=processor,
+        pipeline_class=LayoutLMInferencePipeline,
+        device=device,
+    )
+
+    logits = []
+    predictions = []
+    all_output_df = pd.DataFrame()
+    for logit, pred, output_df in pipe(_get_data(dataset)):
+        logits.append(logit)
+        predictions.append(pred)
+        all_output_df = pd.concat([all_output_df, output_df])
+    all_output_df.columns.name = None
+    all_output_df = all_output_df.reset_index(drop=True)
+    all_output_df = clean_extracted_df(all_output_df)
+    all_output_df = all_output_df[["id", "subsidiary", "loc", "own_per"]]
+    return logits, predictions, all_output_df
+
+
+class LayoutLMInferencePipeline(Pipeline):
+    """Pipeline for performing inference with fine-tuned LayoutLM."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize LayoutLMInferencePipeline."""
+        super().__init__(*args, **kwargs)
+
+    def _sanitize_parameters(self, **kwargs):
+        preprocess_kwargs = {}
+        if "maybe_arg" in kwargs:
+            preprocess_kwargs["maybe_arg"] = kwargs["maybe_arg"]
+        return preprocess_kwargs, {}, {}
+
+    def preprocess(self, example):
+        """Encode and tokenize model inputs."""
+        image = example["image"]
+        words = example["tokens"]
+        boxes = example["bboxes"]
+        encoding = self.tokenizer(
+            image,
+            words,
+            boxes=boxes,
+            return_tensors="pt",
+            truncation=True,
+            padding="max_length",
+            max_length=512,  # this is the maximum max_length
+            stride=128,
+            return_offsets_mapping=True,
+            return_overflowing_tokens=True,
+        )
+        model_inputs = {}
+        model_inputs["raw_encoding"] = encoding.copy()
+        model_inputs["example"] = example
+        model_inputs["offset_mapping"] = encoding.pop("offset_mapping")
+        model_inputs["sample_mapping"] = encoding.pop("overflow_to_sample_mapping")
+        # TODO: do we actually need to make these into ints?
+        encoding["input_ids"] = encoding["input_ids"].to(torch.int64)
+        encoding["attention_mask"] = encoding["attention_mask"].to(torch.int64)
+        encoding["bbox"] = encoding["bbox"].to(torch.int64)
+        encoding["pixel_values"] = torch.stack(encoding["pixel_values"])
+        model_inputs["encoding"] = encoding
+        return model_inputs
+
+    def _forward(self, model_inputs):
+        encoding = model_inputs["encoding"]
+        if torch.cuda.is_available():
+            encoding.to("cuda")
```

Comment:
When I run the integration tests I'm getting the following error here: `AttributeError: 'UserDict' object has no attribute 'to'`. Do you know what type you expect here?

Reply:
This should be a Hugging Face type I'm pretty sure. I'll look into this. I've gotten that error before but I don't think I get it here.

Reply:
The type I'm seeing is `transformers.tokenization_utils_base.BatchEncoding`

Reply:
I turned `encoding` back into a `BatchEncoding` in this `_forward` function, which I think should fix the issue.

Reply:
That worked for me!